### ISSUE TITLE
torch.gather behavior changed from 1.5.1 to master

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#41532]
## 🐛 Bug

The documentation states that for `torch.gather(input, dim, index)`, the `index` tensor must have the same size in all dimensions as `input`, except for dimension `dim`. This check is respected in PyTorch 1.5.1, but not on master.

## To Reproduce

On 1.5.1: 
```
>>> t = torch.tensor([[1,2],[3,4]])
>>> index = torch.tensor([[0]])
>>> torch.gather(t, 1, index)
RuntimeError: Size does not match at dimension 0 get 2 vs 1
```

On master:
```
t = torch.tensor([[1,2],[3,4]])
index = torch.tensor([[0]])
torch.gather(t, 1, index)
```

## Expected behavior

I'm not sure. Either we should update the documentation or add an error check.

**Comment by gchanan** [comment]
bisect points here: https://github.com/pytorch/pytorch/pull/34690

**Comment by gchanan** [comment]
that change seems okay, because the relationship is still well defined, i.e. for dim==1 we have
`out[i][j] = input[i][index[i][j]]`
so given that `index.shape == out.shape` and that the shape out the output is smaller in the non-dim dimensions that's a well defined relationship.

**Comment by gchanan** [comment]
this seems strange, though:
```
torch.gather(torch.randn(3,3), 0, torch.tensor([0]))
tensor([-0.1960])
```

so you'd have...
`out[i] = input[index[i]]` but `input[index[i]` is not a value.

**Comment by gchanan** [comment]
should probably do similar checks for scatter.

**Comment by gchanan** [comment]
other strangeness:
```
>>> x
tensor([[-1.1485, -0.3051, -0.2253],
        [ 0.4836, -0.7794, -1.8069],
        [-0.5331,  1.0695, -2.0397]])
>>> torch.gather(x, 0, torch.tensor([[[0]]]))
tensor([[[-1.1485]]])
>>> torch.gather(x, 1, torch.tensor([[[0]]]))
tensor([[[-1.1485]]])
>>> torch.gather(x, 2, torch.tensor([[[0]]]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: Dimension out of range (expected to be in range of [-2, 1], but got 2)
```

if we have:
```
out(0,0,0) = input(index(0,0,0), 0, 0) OR
out(0,0,0) = input(0, index(0,0,0), 0) OR
out(0,0,0) = input(0, 0, index(0,0,0)))
```

why is dim=2 not allowed?  It's all nonsense anyway.

**Comment by gchanan** [comment]
from bisecting, it looks like this starts being accepted in https://github.com/pytorch/pytorch/pull/36875/.  Probably because it's an optimization that considered skipping over the TORCH_INTERNAL_ASSERT to be okay.

**Comment by ngimel** [comment]
#40359 is also related. @nikitaved can you please tighten the shape checks so that requirements for index dimensions are enforced?

**Comment by nikitaved** [comment]
The current check for gather makes sure that `src.size(i) == index.size(i) i != dim` and `src.size(dim) <= index.size(dim)`.
Shall we leave or change this behavior? What is missing in the current checks is the dimensionality match check... This I will add.

**Comment by nikitaved** [comment]
> torch.gather(x, 2, torch.tensor([[[0]]]))

The tensor `x` seems to be `2-dim` in the example. This one works:
```python
In [14]: x = torch.tensor([[[-1.1485, -0.3051, -0.2253],
    ...:         [ 0.4836, -0.7794, -1.8069],
    ...:                 [-0.5331,  1.0695, -2.0397]]])

In [15]: torch.gather(x, 2, torch.tensor([[[0]]]))
Out[15]: tensor([[[-1.1485]]])

```

But the check for valid `dim` value is missing, yes, so I will add that.

**Comment by nikitaved** [comment]
> other strangeness:
> 
> ```
> >>> x
> tensor([[-1.1485, -0.3051, -0.2253],
>         [ 0.4836, -0.7794, -1.8069],
>         [-0.5331,  1.0695, -2.0397]])
> >>> torch.gather(x, 0, torch.tensor([[[0]]]))
> tensor([[[-1.1485]]])
> >>> torch.gather(x, 1, torch.tensor([[[0]]]))
> tensor([[[-1.1485]]])
> >>> torch.gather(x, 2, torch.tensor([[[0]]]))
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> IndexError: Dimension out of range (expected to be in range of [-2, 1], but got 2)
> ```
> 
> if we have:
> 
> ```
> out(0,0,0) = input(index(0,0,0), 0, 0) OR
> out(0,0,0) = input(0, index(0,0,0), 0) OR
> out(0,0,0) = input(0, 0, index(0,0,0)))
> ```
> 
> why is dim=2 not allowed? it doesn't follow indexing semantics anyway.

This example has x as a 2-dim tensor, while the index is a 3-dim tensor. So it does not work for me:
```python
In [22]: x = torch.tensor([[-1.1485, -0.3051, -0.2253],
    ...:         [ 0.4836, -0.7794, -1.8069],
    ...:                 [-0.5331,  1.0695, -2.0397]])

In [23]: torch.gather(x, 0, torch.tensor([[[0]]]))
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-23-dd33d8016b64> in <module>
----> 1 torch.gather(x, 0, torch.tensor([[[0]]]))

IndexError: Dimension out of range (expected to be in range of [-2, 1], but got 2)

```

This one does work:
```python
In [24]: x = torch.tensor([[[-1.1485, -0.3051, -0.2253],
    ...:         [ 0.4836, -0.7794, -1.8069],
    ...:                 [-0.5331,  1.0695, -2.0397]]])

In [25]: torch.gather(x, 2, torch.tensor([[[0]]]))
Out[25]: tensor([[[-1.1485]]])

```

**Comment by nikitaved** [comment]
I have included the checks for equal dimensions. The gather behavior on inputs with `index.size(i) <= input.size(i), i != dim` is a well-defined behavior which is similar to what scatter is doing. Shall we update the docs?

**Comment by gchanan** [comment]
@nikitaved where you have included the checks for equal dimensions?

**Comment by nikitaved** [comment]
Ah, sorry, here it goes: [#41672](https://github.com/pytorch/pytorch/pull/41672).

**Comment by gchanan** [comment]
I think I did the previous `gather` examples from the commits I linked instead of the release branch.

```
>>> x
tensor([[-0.9640, -1.2366, -1.1208],
        [ 1.8341,  0.6243, -0.5042],
        [ 0.0528,  0.2580,  0.2960]])
# we don't check the dimensionality matches, but we do check it's not larger
>>> torch.gather(x, 0, torch.tensor(0))
tensor(-0.9640)
>>>torch.gather(x, 0, torch.tensor([0]))
tensor([-0.9640])
>>> torch.gather(x, 0, torch.tensor([[0]]))
tensor([[-0.9640]])
>>> torch.gather(x, 0, torch.tensor([[[0]]]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: Dimension out of range (expected to be in range of [-2, 1], but got 2)
```

**Comment by gchanan** [comment]
Fixed by #41672.
